lib/coverband/adapters/base.rb

```diff
@@ -4,36 +4,78 @@ module Coverband
   module Adapters
     class Base
       def initialize
-        raise 'abstract'
+        @file_hash_cache = {}
       end
 
       def clear!
         raise 'abstract'
       end
 
-      def save_report(_report)
-        raise 'abstract'
+      # Note: This could lead to slight race on redis
+      # where multiple processes pull the old coverage and add to it then push
+      # the Coverband 2 had the same issue,
+      # and the tradeoff has always been acceptable
+      def save_report(report)
+        data = report.dup
+        merge_reports(data, get_full_report)
+        save_coverage(data)
       end
 
       def coverage
-        raise 'abstract'
+        simple_report(get_full_report)
       end
 
       def covered_files
+        coverage.keys || []
+      end
+
+      def covered_lines_for_file(file)
+        coverage[file] || []
+      end
+
+      protected
+
+      def save_coverage
         raise 'abstract'
       end
 
-      def covered_lines_for_file(_file)
+      def get_report
         raise 'abstract'
       end
 
-      protected
+      def get_full_report
+        get_report
+      end
+
+      def file_hash(file)
+        if @file_hash_cache[file]
+          @file_hash_cache[file]
+        else
+          @file_hash_cache[file] = Digest::MD5.file(file).hexdigest
+        end
+      end
+
+      def expand_report(report)
+        report.each_pair do |key, line_data|
+          extended_data = {
+            'first_updated_at' => Time.now.to_i,
+            'last_updated_at' => Time.now.to_i,
+            'file_hash' => file_hash(key),
+            'data' => line_data
+          }
+          report[key] = extended_data
+        end
+        report
```

Comment:
Think you can leave off this line as each_pair returns the modified hash. Did a little test to double check:
```
irb(main):002:0> hash = {a:1, b:2}
=> {:a=>1, :b=>2}
irb(main):003:0> hash.each_pair{|k,v| hash[k] = v+1}
=> {:a=>2, :b=>3}
```